drgn_tools/slabinfo.py

```diff
@@ -0,0 +1,310 @@
+    output.append(
+        [
+            ad,
+            int(s.object_size),
+            inuse,
+            total_inuse,
+            total_slabs,
+            f"{ssize}k",
+            s.name.string_().decode("utf-8"),
+        ]
+    )
+
+
+def freelist_dereference_swab(
+    prog: drgn.Program, slab_cache: drgn.Object, ptr_addr: int
+) -> int:
+    """
+    Freelist traversal with swab
+    """
+    freelist_random = slab_cache.random.value_()
+    ulong_size = sizeof(prog.type("unsigned long"))
+    return (
+        prog.read_word(ptr_addr)
+        ^ freelist_random
+        ^ int.from_bytes(ptr_addr.to_bytes(ulong_size, "little"), "big")
+    )
+
+
+def freelist_dereference_noswab(
+    prog: drgn.Program, slab_cache: drgn.Object, ptr_addr: int
+) -> int:
+    """
+    Freelist traversal with no swab
+    """
+    freelist_random = slab_cache.random.value_()
+    return prog.read_word(ptr_addr) ^ freelist_random ^ ptr_addr
+
+
+def try_hardened_freelist_dereference(
+    prog: drgn.Program, slab_cache: drgn.Object, ptr_addr: int
+) -> int:
+    """
+    Segregate the free list traverse based on CONFIG_SLAB_FREELIST_HARDENED
+    """
+    try:
+        slab_cache.random.value_()
+
+    except AttributeError:
+        slab_freelist_hardened = 0
+        needswab = 0
+
+    else:
+        slab_freelist_hardened = 1
+        needswab = 1
+
+    if slab_freelist_hardened and needswab:
+        result = freelist_dereference_swab(prog, slab_cache, ptr_addr)
+    if not slab_freelist_hardened and not needswab:
+        try:
+            result = prog.read_word(ptr_addr)
+        except (AttributeError, drgn.ObjectAbsentError):
+            result = 0
+
+    return result
+
+
+def slub_get_freelist(
+    prog: drgn.Program,
+    slab_cache: drgn.Object,
+    freelist: drgn.Object,
+    freelist_set: Set[int],
+) -> None:
+    """
+    Main function to traverse through the freelist data of kmem_cache_cpu
+    """
+    ptr = freelist.value_()
+    while ptr:
+        if ptr & PAGE_MAPPING_ANON:
+            break
+        freelist_set.add(ptr)
+        ptr = try_hardened_freelist_dereference(
+            prog, slab_cache, (ptr + slab_cache.offset.value_())
+        )
+
+
+def slub_get_cpu_freelist(
+    prog: drgn.Program, slab_cache: drgn.Object, cpu_freelist: drgn.Object
+) -> set:
+    """
+    Collect freelist data from kmem_cache_cpu and
+    return the set of free objects
+    """
+    cpu_free_set: Set[int] = set()
+    slub_get_freelist(prog, slab_cache, cpu_freelist, cpu_free_set)
+    return cpu_free_set
+
+
+def slub_get_cpu_freelist_cnt(
+    prog: drgn.Program, slab_cache: drgn.Object, cpu_freelist: drgn.Object
+) -> int:
+    """
+    Collect freelist data from kmem_cache_cpu and
+    return the count of free objects
+    """
+    cpu_free_set_rc = slub_get_cpu_freelist(prog, slab_cache, cpu_freelist)
+    return len(cpu_free_set_rc)
+
+
+def slub_partial_free(
+    prog: drgn.Program, slab_cache: drgn.Object, slab: drgn.Object
```

Comment:
you can get "prog" from slab_cache.prog_ or slab.prog_. Also partial slabs can be both in per-cpu and per-node lists. Although the description indicates that it is using per-cpu slab list (i.e from kmem_cache_cpu) the name of the function itself should reflect that we are using per-cpu and not per node partial slab list here. I don't see slab_cache being used in this function so you can remove it. Or since this is supposed to find number of objects on per-cpu partial slabs for a give cache and cpu, you can avoid passing slab and just pass the slab_cache and cpu number